contrib/raftexample/raft.go

```diff
@@ -0,0 +1,246 @@
+// Copyright 2015 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"log"
+	"os"
+	"strconv"
+	"time"
+
+	"encoding/gob"
+	"net/http"
+	"net/url"
+
+	"github.com/coreos/etcd/Godeps/_workspace/src/golang.org/x/net/context"
+	"github.com/coreos/etcd/etcdserver/stats"
+	"github.com/coreos/etcd/pkg/types"
+	"github.com/coreos/etcd/raft"
+	"github.com/coreos/etcd/raft/raftpb"
+	"github.com/coreos/etcd/rafthttp"
+	"github.com/coreos/etcd/wal"
+	"github.com/coreos/etcd/wal/walpb"
+)
+
+type raftKV struct {
+	Key string
+	Val []byte
+}
+
+// A key-value stream backed by raft
+type raftKVNode struct {
+	proposeC <-chan raftKV // proposed messages (k,v)
+	commitC  chan raftKV   // entries committed to log (k,v)
+	errorC   chan error    // errors from raft session
+
+	id    int      // client ID for raft session
+	peers []string // raft peer URLs
+
+	// raft backing for the commit/error channel
+	node        raft.Node
+	raftStorage *raft.MemoryStorage
+	wal         *wal.WAL
+	transport   *rafthttp.Transport
+}
+
+// A raft channel represents a key-value raft workflow over http using channels:
+// * client sends (k,v) change proposals by writing to a proposal channel.
+// * client receives a stream of commits by reading a commit channel.
+//   On initialization, the write ahead log replays into the commit channel; to
+//   avoid exposing stale values, a nil value over the commit channel signals
+//   the replay is complete.
+// * client receives raft errors via an out-of-band error channel.
+func newRaftChan(id int, peers []string, proposeC <-chan raftKV) (<-chan raftKV, <-chan error) {
```

Comment:
The comment style in go like: newRaftchan is something (or does something). Then follow with more details.

```
newRaftChan returns two chans for receiving committed entries and any errors. (...) 
```